Model/ReceiptDataProvider.php

```diff
@@ -549,7 +505,21 @@ protected function addPaymentTransaction(\Magento\Sales\Model\Order $order, $tra
     private function cancelOrderById($orderId): void
     {
         if ($this->gatewayConfig->getCancelOrderOnFailedPayment()) {
-            $this->orderManagementInterface->cancel($orderId);
+            try {
+                $this->orderManagementInterface->cancel($orderId);
+            } catch (\Exception $e) {
+                $this->logger->critical(sprintf(
+                    'Paytrail exception during order cancel: %s,\n error trace: %s',
+                    $e->getMessage(),
+                    $e->getTraceAsString()
+                ));
+
+                // Mask and throw end-user friendly exception
+                throw new CheckoutException(__(
+                    'Error while cancelling order. Please contact customer support with order id: %id to release discount coupons.',
```

Comment:
This is order cancellation, does it contain always discount coupons or is that more or less general message that if you had discount coupons used then contact support to release those?

Reply:
Up to this date the only time this issue has been encountered has been when order has contained discount coupons and Magento's background process has not managed to mark the coupon as used before the order was cancelled. But the nature of this message is general "If you had coupons, they need to be released manually"